### PR TITLE
Update RustCrypto crates

### PR DESCRIPTION
**Pull request by paolobarbolini**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -91,11 +104,19 @@ where
...
+        + Default
+        + Clone,
+    <H::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
+    Le<<H::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
 {
     // Hash the secret and the time together.
-    let mut mac: Hmac<H> = Hmac::new_from_slice(secret).unwrap();
-    mac.update(&to_bytes(time / step));
+    let mut mac = <Hmac<H> as Mac>::new_from_slice(secret).unwrap();
+    <Hmac<H> as Update>::update(&mut mac, &to_bytes(time / step));
```
Comment (fosskers) [inline src/lib.rs:119]: Is `update` no longer a method?

File: src/lib.rs [hunk 2]
```diff
@@ -68,7 +73,15 @@ pub const DEFAULT_DIGITS: u32 = 8;
...
-    H: Update + BlockInput + Reset + FixedOutputDirty + Clone + Default,
+    H: Update + FixedOutput + CoreProxy,
+    H::Core: HashMarker
+        + UpdateCore
+        + FixedOutputCore
+        + BufferKindUser<BufferKind = Eager>
+        + Default
+        + Clone,
+    <H::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
+    Le<<H::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
```
[the remark this answers is not in the document]
Comment (paolobarbolini) [reply src/lib.rs:84]: I mainly copy-pasted them from `Hmac` https://docs.rs/hmac/latest/src/hmac/optim.rs.html#20-36

File: src/lib.rs (hunk 1 above)
Comment (paolobarbolini) [reply src/lib.rs:119]: It is still a method, but both the `Update` and the `UpdateCore` traits implement `update`, so I had to let the compiler know which of the two implementations to use

**Comment by BlackDex** [comment]
Nice, thanks @paolobarbolini, I was looking through the dependencies of Vaultwarden, and saw an easy update on duplicate crates here. This saves me creating an PR ;).

And @fosskers thanks for this crate, works really nice with Vaultwarden, would be great to have an updated release!

**Comment by paolobarbolini** [comment]
> I just released `1.1.0`. Thanks for the PR!

Shouldn't it be `2.0.0`? I know you re-export the RustCrypto hash functions implementations that can be used for TOTP, but nothing prevents users from not using the re-export and importing the implementations themselves, so this is a breaking change.

**Comment by fosskers** [comment]
Could a break actually occur? Even if they imported say `Sha256` from the underlying lib themselves and passed that in, it should still work without issue.

**Comment by paolobarbolini** [comment]
> Could a break actually occur? Even if they imported say `Sha256` from the underlying lib themselves and passed that in, it should still work without issue.

Nope, because the traits are from a different version of the `digest` crate, so they won't match

**Comment by fosskers** [comment]
deps

`totp-lite` depends on `digest 0.10`, as do the three crypto libs, `hmac`, `sha2`, and `sha-1`. I suppose you're implying there would be breaks if the user was manually depending on `digest 0.9` (already two years old).

**Comment by paolobarbolini** [comment]
> I suppose you're implying there would be breaks if the user was manually depending on `digest 0.9` (already two years old)?

It might be improbable, maybe nobody will ever complain about this release, but upgrading a public dependency in a backwards incompatible way is a breaking change. It's annoying [sometimes](https://github.com/launchbadge/sqlx/pull/1505#issuecomment-1027664202) but it gives strong semver guarantees

**Comment by fosskers** [comment]
Well I wouldn't want to break semver. So be it, I just released `2.0.0` and yanked `1.1.0` off crates.io

**Pull request merged**